### PR TITLE
feat: add AntV diagrams + Leaflet maps

### PR DESCRIPTION
**Pull request by OchnikBartek**
## Summary

Adds AntV diagram sidecar + Leaflet map rendering to the generated template, with reliability fixes for MCP tooling across all 6 AI frameworks.

## Changes

- Add `enable_antv_charts` cookiecutter variable gating AntV MCP sidecar, diagram tools, and Leaflet map rendering
- Add `create_map` tool with `MapMarker` typed schema — prevents empty-marker validation errors from weaker models (GPT-4.5 etc.)
- Wire `create_map` into all 6 framework assistants with typed signatures for JSON schema fidelity
- Add render-once prompt nudge for maps, charts, and AntV diagrams (fixes Opus re-render duplicate bug)
- Fix MCP connection leak in CrewAI: `get_antv_crewai_tools` memoizes the started adapter so it's started once per process, not once per request
- Fix per-request event-loop blocking in LangChain/LangGraph/DeepAgents: memoize tool discovery so `_run_sync` only blocks on first request
- Fix antvis-chart Docker healthcheck: plain `wget` rejected by streamable-HTTP endpoint → replaced with `node` one-liner
- Add coordinate bounds validation to `MapSpec._validate_center` (matches per-marker validation)
- Remove dead `parse_map_spec` export
- Add `react-leaflet` map renderer in frontend (`MapMessage`, `MapLeaflet` components)
- Add AntV image renderer in `tool-call-card.tsx`
- Add matrix test entries for `pydantic_ai`, `langchain`, `crewai` with `enable_antv_charts` and `enable_charts`
- Add `TestGeneratedTemplateAntvCharts` content tests verifying typed markers, bounds check, cache variables, and absence of removed dead code

## Testing

- [x] Matrix tests: generated pydantic_ai / langchain / crewai / pydantic_deep with `--antv-charts`; all render, compile, pass ruff
- [x] `TestGeneratedTemplateAntvCharts` content tests: `MapMarker` present, bounds check present, `parse_map_spec` absent, `list[MapMarker]` in assistant, module-level cache in antv_chart.py
- [x] Typed-marker fix verified via LangChain smoke test (dict markers coerced, empty markers rejected)
- [ ] All tests pass locally: `make test`
- [x] Linting passes: `make lint`
- [ ] Type checking passes: `make typecheck`
- [x] Generated projects still work: tested with at least one preset

### CODE REVIEW

**Review by bll-kacper-wlodarczyk — COMMENTED** [review]
Solid, well-documented feature — the per-framework MCP wiring, graceful degradation, typed `MapMarker` schema, and the dedicated content tests are all nicely done. A few things worth addressing before merge: a doc/compose mismatch around the production compose, publishing the unauthenticated MCP port to the host in prod, and two edge cases in the tool-discovery memoization (thread-safety race + permanent caching of a transient failure). Inline comments below.

**Review by DEENUU1 — COMMENTED** [review]
Solid, well-documented feature — the per-framework MCP wiring, graceful degradation, typed `MapMarker` schema, and the dedicated content tests are all nicely done. A few things worth addressing before merge: a doc/compose mismatch around the production compose, publishing the unauthenticated MCP port to the host in prod, and two edge cases in the tool-discovery memoization (thread-safety race + permanent caching of a transient failure). Inline comments below.

File: template/VARIABLES.md [hunk 1]
```diff
@@ -286,6 +286,7 @@ These variables are set automatically by the generator.
 | `web_fetch_tool` | bool | `false` | Computed: portable `fetch_url` tool is generated | `enable_web_fetch` and framework is LangChain/LangGraph/CrewAI/DeepAgents |
 | `enable_charts` | bool | `false` | Enable the chart-generation tool (line/bar/pie/area/scatter); interactive in web chat, PNG on Slack/Telegram | Requires an AI framework |
 | `charts_channel_png` | bool | `false` | Computed: render charts to PNG for messaging channels | `enable_charts` and (`use_slack` or `use_telegram`) |
+| `enable_antv_charts` | bool | `false` | Adds the interactive `create_map` tool (Leaflet/OpenStreetMap, works out of the box) plus AntV advanced-diagram tools (flowchart, mind-map, org-chart, sankey, ...) via an `mcp-server-chart` Docker sidecar. Wired into all 5 frameworks. The AntV diagrams need the sidecar running (`docker compose --profile antv up -d`) and `ENABLE_ANTV_CHARTS=true` at runtime; maps work without it. Not wired into the production compose (internal network blocks AntV's render backend — self-host GPT-Vis-SSR for prod) | Requires an AI framework |
```
Comment (DEENUU1) [inline template/VARIABLES.md]: This says the sidecar is **not wired into the production compose**, but `docker-compose.yml` (the prod file) *does* add the `antvis-chart` service — it's just gated behind the `antv` profile, so it doesn't start by default. The statement is misleading: it is wired in, just opt-in. Consider rewording to something like "present in the prod compose but profile-gated; not recommended for prod (...)" so the docs match what the templates actually generate.

File: template/{{cookiecutter.project_slug}}/docker-compose.yml [hunk 2]
```diff
@@ -447,6 +447,44 @@ services:
...
+      --transport streamable --host 0.0.0.0 --port 1122 --endpoint /mcp
+    environment:
+      # Empty = AntV's public rendering backend. Point at a self-hosted
+      # GPT-Vis-SSR for private deployments (data never leaves your infra).
+      - VIS_REQUEST_SERVER=${ANTV_VIS_REQUEST_SERVER:-}
+      # Default drops basic charts (covered by create_chart/Recharts) and the
+      # China-only map tools (covered by create_map/Leaflet). Override via ANTV_DISABLED_TOOLS.
+      - DISABLED_TOOLS=${ANTV_DISABLED_TOOLS:-generate_area_chart,generate_bar_chart,generate_column_chart,generate_line_chart,generate_pie_chart,generate_scatter_chart,generate_district_map,generate_pin_map,generate_path_map}
+    ports:
+      - "1122:1122"
```
Comment (DEENUU1) [inline template/{{cookiecutter.project_slug}}/docker-compose.yml]: The backend talks to the sidecar over the internal `backend` network (`ANTV_MCP_URL=http://antvis-chart:1122/mcp`), so this host port mapping isn't needed for the feature to work. In the **production** compose, `1122:1122` publishes an unauthenticated MCP + render endpoint on the host. I'd drop the `ports` block here and keep it only in `docker-compose.dev.yml` if host access is wanted for debugging.

File: template/{{cookiecutter.project_slug}}/backend/app/agents/tools/antv_chart.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    if not settings.ENABLE_ANTV_CHARTS:
+        return []
+    if _antv_crewai_tools is not None:
+        return _antv_crewai_tools
+    try:
+        from crewai_tools import MCPServerAdapter
+
+        server_params = {"url": settings.ANTV_MCP_URL, "transport": "streamable-http"}
+        adapter = MCPServerAdapter(server_params)
+        adapter.start()
```
Comment (DEENUU1) [inline template/{{cookiecutter.project_slug}}/backend/app/agents/tools/antv_chart.py]: The memoization here isn't concurrency-safe. FastAPI runs sync request handlers in a threadpool, so two requests racing on the very first call can both pass the `if _antv_crewai_tools is not None` check above and each reach `adapter.start()` — re-introducing the per-request adapter leak this function is meant to fix. A module-level `threading.Lock` around the check-and-start (double-checked) would close the race, or start the adapter once at app startup instead of lazily.

File: template/{{cookiecutter.project_slug}}/backend/app/agents/tools/antv_chart.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        async def _load() -> list[Any]:
+            client = MultiServerMCPClient(
+                {"antv": {"url": settings.ANTV_MCP_URL, "transport": "streamable_http"}}
+            )
+            return await client.get_tools()
+
+        _antv_langchain_tools = _run_sync(_load())
+    except Exception as exc:
+        logger.warning("AntV MCP tools unavailable, continuing without them: %s", exc)
+        _antv_langchain_tools = []
```
Comment (DEENUU1) [inline template/{{cookiecutter.project_slug}}/backend/app/agents/tools/antv_chart.py]: On a transient failure this caches `[]` **permanently**: `[]` is `not None`, so once the first call fails (e.g. the sidecar's 20s `start_period` hasn't elapsed when the first request builds the agent), every later call returns the cached empty list and AntV stays disabled until the process restarts — even after the sidecar becomes healthy. Consider memoizing only on success (leave the global as `None` in the `except` so the next request retries). Same pattern applies to `get_antv_crewai_tools`.

**Review by DEENUU1 — APPROVED** [review]
lgtm
